### PR TITLE
SONARIAC-681 Rule S6437: Add secret generation detection for ssh-keygen

### PR DESCRIPTION
**Pull request by jonas-wielage-sonarsource**

### CODE REVIEW

**Comment by jonas-wielage-sonarsource** [comment]
I just wanted to have some feedback for this PR.
Code Smells and coverage are something I will fix beforehand.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java [hunk 1]
```diff
@@ -78,44 +82,149 @@ public List<Command> search(List<ArgumentResolution> resolvedArguments) {
...
   private List<ArgumentResolution> fullMatch(Deque<ArgumentResolution> argumentStack) {
-    List<ArgumentResolution> commandArguments = new ArrayList<>();
-    Deque<CommandPredicate> predicateStack = new LinkedList<>(predicates);
-    while (!predicateStack.isEmpty()) {
-      CommandPredicate currentPredicate = predicateStack.pollFirst();
-      ArgumentResolution resolution = argumentStack.pollFirst();
+    CommandDetectionStatus detectionStatus = new CommandDetectionStatus(argumentStack, new LinkedList<>(predicates), new ArrayList<>());
+    while (!detectionStatus.predicateStack.isEmpty()) {
+
+      detectionStatus.setCurrentPredicate(detectionStatus.predicateStack.pollFirst());
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java:90]: This method does not need a parameter as the only provided argument is of the object itself

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java [hunk 2]
```diff
@@ -78,44 +82,149 @@ public List<Command> search(List<ArgumentResolution> resolvedArguments) {
...
   // Cognitive Complexity of methods should not be too high
   @SuppressWarnings("java:S3776")
+  // TODO: Rework commentary
   private List<ArgumentResolution> fullMatch(Deque<ArgumentResolution> argumentStack) {
-    List<ArgumentResolution> commandArguments = new ArrayList<>();
-    Deque<CommandPredicate> predicateStack = new LinkedList<>(predicates);
-    while (!predicateStack.isEmpty()) {
-      CommandPredicate currentPredicate = predicateStack.pollFirst();
-      ArgumentResolution resolution = argumentStack.pollFirst();
+    CommandDetectionStatus detectionStatus = new CommandDetectionStatus(argumentStack, new LinkedList<>(predicates), new ArrayList<>());
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java:87]: As the constructor is only used once, providing the stack and the predicates should be enough. The empty array as well as the List to LinkedList conversion can be handled in the constructor itself

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java [hunk 3]
```diff
@@ -78,44 +82,149 @@ public List<Command> search(List<ArgumentResolution> resolvedArguments) {
...
+    while (!detectionStatus.predicateStack.isEmpty()) {
+
+      detectionStatus.setCurrentPredicate(detectionStatus.predicateStack.pollFirst());
+      // resolution is removed from stack during match-methods, here it is peeked to see if it is null or UNRESOLVED
+      ArgumentResolution resolution = detectionStatus.argumentStack.peekFirst();
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java:92]: Implement a method instead of a field access

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java [hunk 4]
```diff
@@ -78,44 +82,149 @@ public List<Command> search(List<ArgumentResolution> resolvedArguments) {
...
 
       // Stop argument detection when argument list is empty
       if (resolution == null) {
-        return remainingPredictsAreOptional(currentPredicate, predicateStack) ? commandArguments : Collections.emptyList();
+        return remainingPredictsAreOptional(detectionStatus.getCurrentPredicate(), detectionStatus.predicateStack) ? detectionStatus.commandArguments : Collections.emptyList();
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java:96]: Do not mix getters and direct field access. I would recommend a getter setup.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java [hunk 5]
```diff
@@ -78,44 +82,149 @@ public List<Command> search(List<ArgumentResolution> resolvedArguments) {
...
-        // Re-add predicate to stack to be reevaluated on the next argument
-        if (currentPredicate.is(ZERO_OR_MORE)) {
-          predicateStack.addFirst(currentPredicate);
-        }
-        // Add matched argument to command
-        commandArguments.add(resolution);
-      } else if (currentPredicate.is(OPTIONAL, ZERO_OR_MORE, NO_MATCH)) {
-        // Re-add argument to be evaluated by the next predicate
-        argumentStack.addFirst(resolution);
+      if (detectionStatus.getCurrentPredicate() instanceof SingularPredicate) {
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java:105]: As you call the methods 7 times inside the loop maybe store its value in a variable as it is not expected to return different values withing a single loop iteration.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java [hunk 6]
```diff
@@ -78,44 +82,149 @@ public List<Command> search(List<ArgumentResolution> resolvedArguments) {
...
+      // above method should return empty
+      detectionStatus.setStatus(CommandDetectionStatus.Status.FOUND_NOTHING);
+      return;
+    }
+    detectionStatus.setStatus(CommandDetectionStatus.Status.CONTINUE);
+  }
+
+  private void matchPredicate(OptionPredicate optionPredicate, CommandDetectionStatus detectionStatus) {
+    matchPredicate(optionPredicate.flagPredicate, detectionStatus);
+    if (detectionStatus.is(CommandDetectionStatus.Status.ABORT, CommandDetectionStatus.Status.FOUND_NOTHING) || optionPredicate.withoutValue()) {
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java:161]: I would replace `optionPredicate.withoutValue()` with `optionPredicate.valuePredicate == null` as this helps analyzers to see that 166 is not null

**Review by nils-werner-sonarsource — COMMENTED** [review]
A lot of changes! Nice approach.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandPredicate.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.checks.utils;
+
+public interface CommandPredicate {
+
+  enum Type {
+    MATCH,
+    NO_MATCH,
+    OPTIONAL,
```
[the remark this answers is not in the document]
Comment (rudy-regazzoni-sonarsource) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandPredicate.java:27]: In the end I did not remove `OPTIONAL` : at the time I wanted to merge the PR which was supposed to remove it, @jonas-wielage-sonarsource introduced a PR for `ClearTextProtocolDowngradeCheck` which was using `withOptional()` a lot, so I kept it.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+      }
+
+      // choose match-method depending on implementation of currentPredicate
+      if (currentPredicate instanceof SingularPredicate) {
+        matchPredicate((SingularPredicate) currentPredicate);
+      } else if (currentPredicate instanceof OptionPredicate) {
+        matchPredicate((OptionPredicate) currentPredicate);
+      } else if (currentPredicate instanceof MultipleUnorderedOptionsPredicate) {
+        matchPredicate((MultipleUnorderedOptionsPredicate) currentPredicate);
+      }
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java:92]: Seems a bit strange to me, I understand that we need a different logic to be applied depending of the predicate object, but this goes a bit against the OOP concept.

It would be great to be able to extract the specific logics into the related predicate class.

My default reflex would be to have the interface `CommandPredicate` define a method `match(...)` from which each inheriting class `SingularPredicate`/`OptionPredicate`/`MultipleUnorderedOptionsPredicate` would have an implementation. This implementation would be the current code you have in the different `matchPredicate()` method.
But this cannot be that simple since those depends heavily depend of the `CommandMatcher` context.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+    this.status = status;
+  }
+
+  private ArgumentResolution pollFirstArgumentResolution() {
+    return argumentStack.pollFirst();
+  }
+
+  private ArgumentResolution peekFirstArgumentResolution() {
+    return argumentStack.peekFirst();
+  }
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java:281]: Not totally convinced that those method are necessary ^^ I don't have a strong opinion on this, feel free to keep them if you think they bring clarity.

**Review by rudy-regazzoni-sonarsource — COMMENTED** [review]
There is a lot of work there !
We are moving toward what we wanted to avoid : having a really complex `CommandDetector` class ^^' The previous version indeed had strong limitation which were preventing from implementing this ticket properly.
Still because of our current approach with `RUN`/`CMD`, it is hard to parse command without risking misinterpreting.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java (hunk 9 above)
Comment (nils-werner-sonarsource) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java:281]: This was my suggestion, as it might be hard to understand what peekFirst means for argument handling. And I agree that this naming does not bring any big advantages as it duplicates the `argumentStack.peekFirst` wording.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java (hunk 8 above)
Comment (nils-werner-sonarsource) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java:92]: Sounds like a good and clean approach.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java (hunk 9 above)
Comment (jonas-wielage-sonarsource) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java:281]: Yeah, looking over it again, I'm also not convinced about those methods. Because we are using the argumentStack now as an argument in the CommandMatcher class, we can drop the extra method call. Or are there other suggestions?

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java (hunk 8 above)
Comment (jonas-wielage-sonarsource) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java:92]: How would you suggest handling the context of the `CommandMatcher`? The original idea was to extract the methods and arguments to the `CommandMatcher` to not pass 4 arguments in each method call.

Comment (jonas-wielage-sonarsource) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java:92]: One option would be to pass an object to the methods, which wraps the other arguments and the status.
(Essentially an object of the class `CommandMatcher`, which doesn't have the `match()` methods and only works as a wrapper.
But that would be pretty similar to my first approach, where everything was handled in the CommandDetector, and I had an internal class wrapping the arguments. Except that now the `match()` methods would be contained in the implementations of `CommandPredicate` (which I think is a step in the right direction, but not quite ideal)

File: iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/openssl.dockerfile [hunk 10]
```diff
@@ -2,7 +2,7 @@ FROM ubuntu:22.04
 
 # Noncompliant@+1 {{Revoke and change this secret, as it might be compromised.}}
 RUN openssl genrsa -des3 -passout pass:x -out server.pass.key 2048
-#   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/openssl.dockerfile:5]: I would add below test with highlighting :
```
RUN openssl genrsa -des3 -passout pass:x -out server.pass.key 2048 && other command
RUN openssl genrsa -des3 -passout pass:x -out server.pass.key 2048 | other command
RUN openssl genrsa -des3 -passout pass:x -out server.pass.key 2048; other command
```

File: iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/sshKeygen.dockerfile [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+# Noncompliant@+1
+RUN ssh-keygen -N "" -f rsync-key -b 1024 -t dsa
+
+# Noncompliant@+1
+RUN ssh-keygen -N "" -b 1024 -t dsa -f rsync-key
+
+# Noncompliant@+1
+RUN ssh-keygen -N "" -b 1024 -f rsync-key -t dsa
+
+RUN ssh-keygen -N -t dsa -b 1024 -f rsync-key
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/sshKeygen.dockerfile:93]: I would add these test case :
```
RUN ssh-keygen -N ''      -t dsa -b 1024 -f rsync-key
RUN ssh-keygen -N value   -t dsa -b 1024 -f rsync-key
RUN ssh-keygen -N "value" -t dsa -b 1024 -f rsync-key
RUN ssh-keygen -N ""value -t dsa -b 1024 -f rsync-key

ARG EMPTY=""
ARG NOT_EMPTY="value"
RUN ssh-keygen -N $EMPTY     -t dsa -b 1024 -f rsync-key
RUN ssh-keygen -N $NOT_EMPTY -t dsa -b 1024 -f rsync-key
RUN ssh-keygen -N $UNKNOWN   -t dsa -b 1024 -f rsync-key
```

File: iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/sshKeygen.dockerfile [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+RUN ssh-keygen -random1 -N "" -random2 -t dsa -random3 -b 1024 -random4 -f rsync-key -random5
+
+# Noncompliant@+1
+RUN ssh-keygen -random1 random1Value -N "" -random2 -t dsa -random3 =random3Value -b 1024 -random4 -f rsync-key -random5
+
+# Noncompliant@+1
+RUN ssh-keygen -random1 random1Value -N "" -random2 -t dsa -random3 =random3Value -b 1024 -random4 -f rsync-key -random5 random5Option
+
+
+# Noncompliant@+1
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/sshKeygen.dockerfile:17]: I would add this test case (not necessarily here) :
`RUN ssh-keygen -N "" -t dsa -f rsync-key -b 1024 $UNRESOLVED`

File: iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/sshKeygen.dockerfile [hunk 13]
```diff
@@ -0,0 +1,109 @@
+FROM ubuntu:22.04
+
+# Noncompliant@+1 {{Revoke and change this secret, as it might be compromised.}}
+RUN ssh-keygen -N "" -t dsa -b 1024 -f rsync-key
+#   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/sshKeygen.dockerfile:6]: I would add this test case (not necessarily here) :
`RUN ssh-keygen -N "" -t dsa -N "val" -b 1024 -f rsync-key`

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/MultipleUnorderedOptionsPredicate.java [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+    OptionPredicate anyOptionExceptExpectedPredicate = this.calculateAnyOptionMatchingExceptExpected();
+
+    // initial initialization to step into while, could be solved with do-while
+    boolean anythingMatched = true;
+
+    while (anythingMatched && !workingSet.isEmpty()) {
+
+      if (context.areNoArgumentsToHandle()) {
+        if (!context.remainingPredicatesAreOptional()) {
+          context.setStatus(ABORT);
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/MultipleUnorderedOptionsPredicate.java:97]: The code should have more test : if I comment this line, I don't get any failed test.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/MultipleUnorderedOptionsPredicate.java [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+    }
+
+    // calculate success or failure due to size of sets
+    if (fulfilledOptions.size() != expectedMatches) {
+      // not all expectedMatches are fulfilled so overall the MultipleUnorderedOptionsPredicate hasn't matched
+      context.setStatus(ABORT);
+      return;
+    }
+    context.setStatus(PredicateContext.Status.CONTINUE);
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/MultipleUnorderedOptionsPredicate.java:117]: Same here.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/MultipleUnorderedOptionsPredicate.java [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+          context.setStatus(ABORT);
+        }
+        return;
+      }
+
+      anythingMatched = matchExpectedOrAnyOption(context, fulfilledOptions, workingSet, anyOptionExceptExpectedPredicate);
+      if (context.is(ABORT)) {
+        // could also check for FOUND_NO_PREDICATE_MATCH, but renders the boolean anythingMatched useless which is added for better readability of
+        // the control flow
+        return;
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/MultipleUnorderedOptionsPredicate.java:106]: Same here.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/MultipleUnorderedOptionsPredicate.java [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+      }
+
+    }
+
+    // calculate success or failure due to size of sets
+    if (fulfilledOptions.size() != expectedMatches) {
+      // not all expectedMatches are fulfilled so overall the MultipleUnorderedOptionsPredicate hasn't matched
+      context.setStatus(ABORT);
+      return;
+    }
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/MultipleUnorderedOptionsPredicate.java:116]: Same here: even If I comment this whole block, I don't get any failed test.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/OptionPredicate.java [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+      if (context.getCurrentPredicate() instanceof MultipleUnorderedOptionsPredicate && this.valuePredicate.hasType(MATCH)) {
+        context.setStatus(ABORT);
+      }
+      return;
+    }
+    this.valuePredicate.match(context);
+    if (context.is(FOUND_NO_PREDICATE_MATCH)) {
+      // should abort matching process, because the flag matched but the value didn't match, and there shouldn't be two flags that are identical
+      context.setStatus(ABORT);
+    }
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/OptionPredicate.java:78]: Same here: If I comment this, no failed tests.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/OptionPredicate.java [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+      // further action is evaluated by the calling method
+      return;
+    }
+
+    // should not match on empty argumentStack
+    if (context.areNoArgumentsToHandle()) {
+      // in this special case the calling method doesn't handle it properly, so it should be aborted here
+      if (context.getCurrentPredicate() instanceof MultipleUnorderedOptionsPredicate && this.valuePredicate.hasType(MATCH)) {
+        context.setStatus(ABORT);
+      }
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/OptionPredicate.java:71]: Same comment as for the previous file: the code should have more test: if I comment this line, I don't get any failed test.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/SingularPredicate.java [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  @Override
+  public void match(PredicateContext context) {
+    ArgumentResolution resolution = context.getNextArgumentToHandleAndRemoveFromList();
+
+    if (resolution.isUnresolved()) {
+      context.setStatus(ABORT);
+      return;
+    }
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/SingularPredicate.java:61]: Same as before: I can remove this block of code without breaking any test.

**Review by rudy-regazzoni-sonarsource — CHANGES_REQUESTED** [review]
Nice improvement ! The split of the logic seems much better.
I pursued the review more in depth in test code. Either it need more test or it can be simplified a lot: we can remove several instructions and small block of code without breaking any test :D
Please see my comments, I'm also available if you want to discuss this directly.

File: iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/openssl.dockerfile (hunk 10 above)
Comment (jonas-wielage-sonarsource) [reply iac-extensions/docker/src/test/resources/checks/SecretsGenerationCheck/openssl.dockerfile:5]: In this case the other command would also be flagged.

**Review by rudy-regazzoni-sonarsource — APPROVED** [review]
LGTM ! Nice improvement of the `CommandDetector` API
